lib/widget/image_cropper.dart

```diff
@@ -0,0 +1,333 @@
+      'onTap': (funcDefinition) => _controller.onTap =
+          EnsembleAction.fromYaml(funcDefinition, initiator: this),
+      'isRotate': (value) =>
+          _controller.isRotate = Utils.getBool(value, fallback: true),
+      'isMove': (value) =>
+          _controller.isMove = Utils.getBool(value, fallback: true),
+      'isScale': (value) =>
+          _controller.isScale = Utils.getBool(value, fallback: true),
+      'onCropped': (funcDefinition) => _controller.onCropped =
+          EnsembleAction.fromYaml(funcDefinition, initiator: this)
+    };
+  }
+}
+
+mixin CropAction on WidgetState<EnsembleImageCropper> {
+  Future<void> cropImage();
+  void reset();
+  void rotateLeft();
+  void rotateRight();
+}
+
+class EnsembleImageCropperController extends BoxController {
+  CropAction? cropAction;
+  String source = '';
+  String? fit;
+  Color? placeholderColor;
+  EnsembleAction? onTap;
+  double? imageHeight;
+  double? imageWidth;
+
+  double? strokeWidth;
+  double? shapeBorderRadius;
+  double? cropPercentage;
+  String shape = '';
+  Color? strokeColor;
+  bool isRotate = true;
+  bool isMove = true;
+  bool isScale = true;
+  EnsembleAction? onCropped;
+  CustomImageCropController cropController = CustomImageCropController();
+}
+
+class EnsembleImageCropperState extends WidgetState<EnsembleImageCropper>
+    with CropAction {
+  late Widget placeholder;
+  final currentShape = CustomCropShape.Circle;
+
+  @override
+  void initState() {
+    super.initState();
+    placeholder = getPlaceholder();
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    widget.controller.cropAction = this;
+  }
+
+  @override
+  void didUpdateWidget(covariant EnsembleImageCropper oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    widget.controller.cropAction = this;
+  }
+
+  @override
+  Future<void> cropImage() async {
+    final image = await widget.controller.cropController.onCropImage();
+    if (image != null) {
+      io.Directory tempDir = await getTemporaryDirectory();
+      String tempPath = tempDir.path;
+      final filePath =
+          '$tempPath/${DateTime.now().toIso8601String()}-cropped-image.png';
+      final file = io.File(filePath);
+      final byteData = image.bytes.buffer.asUint8List();
+      file.createSync();
+      file.writeAsBytes(byteData.toList());
+
+      ScreenController().executeAction(
+        context,
+        widget._controller.onCropped!,
+        event: EnsembleEvent(
+          widget,
+          data: {'file': file.path},
+        ),
+      );
+    }
+  }
+
+  @override
+  void reset() {
+    widget.controller.cropController.reset();
+  }
+
+  @override
+  void rotateLeft() {
+    widget.controller.cropController
+        .addTransition(CropImageData(angle: -pi / 4));
+  }
+
+  @override
+  void rotateRight() {
+    widget.controller.cropController
+        .addTransition(CropImageData(angle: pi / 4));
+  }
```

Comment:
Why are these just not implemented on the controller directly?

Reply:
I moved crop and rotate, reset related methods to the controller. Can you review it again?